Add support for per-netns config files to pam_usernet

DanielG opened the pull request:
Hi,

I'm trying to use libpam-net to force some users to only be able to communicate through a VPN which works great but I noticed that it doesn't support mounting config files from `/etc/netns/<ns_name>` like iproute2 does. So I added support for it.

This is primarily required to have different DNS setups in the namespaces. For example to prevent DNS leaks in a VPN setup.

To use it create a directory for a user's netns in `/etc/netns/<user>`. Config files placed directly underneath it will get bind mounted over files in `/etc` upon entering the namespace as usual. See ip-netns(8).

FYI: I'm also working on getting libpam-net packaged in Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=909908. If you could please tag a new release soon after merging this so I can get it straight into the first version that would be super :)

DanielG committed:
Add support for per-netns config files to pam_usernet

This is primarily required to support different DNS setups in the namespaces.

To use it create a directory for a user's netns in `/etc/netns/<user>`. Config
files placed directly underneath it will get bind mounted over files in `/etc`
upon entering the namespace as usual. See ip-netns(8).

DanielG committed:
Rename NSDIR to be consistent

DanielG committed:
Remove trailing whitespace

rd235 commented:
Hi,
	just a quick note (busy day)... thank you, the new feature is interesting.

Both libpam-net and iproute2 have been released under GPL2+ so there should not
any problem with the license.

In the new release I would like to:
i. include your support of /etc/netns/<ns_name> configuration files
ii. add some lines with an example in README.md (can you send me some text for this?)
iii. delete config.h (it is a mistake, it is generated by the autotools)
iv. maybe define variable length arrays instead of using PATH_MAX chars for each path
in bind_etc... (when possible I prefer to minimize stack allocation).

happy hacking
renzo

On Sun, Sep 30, 2018 at 03:44:50AM -0700, Daniel Gröber wrote:
> Hi,
> 
> I'm trying to use libpam-net to force some users to only be able to communicate
> through a VPN which works great but I noticed that it doesn't support mounting
> config files from /etc/netns/<ns_name> like iproute2 does. So I added support
> for it.
> 
> This is primarily required to have different DNS setups in the namespaces. For
> example to prevent DNS leaks in a VPN setup.
> 
> To use it create a directory for a user's netns in /etc/netns/<user>. Config
> files placed directly underneath it will get bind mounted over files in /etc
> upon entering the namespace as usual. See ip-netns(8).
> 
> FYI: I'm also working on getting libpam-net packaged in Debian: https://
> bugs.debian.org/cgi-bin/bugreport.cgi?bug=909908. If you could please tag a new
> release soon after merging this so I can get it straight into the first version
> that would be super :)

DanielG committed:
Remove use of VLAs

DanielG committed:
Split pam_sm_open_session functionality into a couple of functions

DanielG committed:
Reindent

DanielG committed:
Remove accidentally commited config.h

DanielG committed:
Do some copy editing on README

DanielG committed:
README: Introduce sections for use cases

DanielG committed:
README: Fix mistake in example code

Linux adds the loopback addresses to 'lo' automatically when it goes up, but you
have to up it to allow communication.

DanielG committed:
README: Add /etc/netns/ config example

DanielG commented:
Finally got around to addressing your suggestions.

re lincense: I assume you're asking why I added the copyright line to the GPL header. Technically the GPL requires you retain existing copyright notices when copying code verbatim. It happens that the code I used as a template doesn't have any explicit copyright notices. However I think it's just good form to disclose the fact that some code was taken from iproute2. It's not as if doing so is going to hurt :)

re PATH_MAX: I don't particularily like VLAs, it is my understanding that they are considered a really bad idea especially in security sensitive code such as this. I've instead change the code to use `asprintf`. This function ensures destination buffers are always the right size based on its inputs which makes it much easier to not misuse than the other `sprintf` variants.

I've also added an example on how to use this new feature to the README like you asked.

rd235 commented:
I have merged your branch.

I have successfully tested the new feature here.
If you could kindly confirm that the current version is okay, I'll add the v.0.2 tag
and then everything is ready for the Debian packetization.

Thank you. ciao
	renzo

On Thu, Oct 04, 2018 at 11:23:28AM -0700, Daniel Gröber wrote:
> Finally got around to addressing your suggestions.
